components/ei-analytics/org.wso2.analytics.solutions.ei.analytics/widgets/EIAnalyticsMessageTable/src/EIAnalyticsMessageTable.jsx

```diff
@@ -90,27 +89,23 @@ class EIAnalyticsMessageTable extends Widget {
         this.handleStats = this.handleStats.bind(this);
         this.handleGraphUpdate = this.handleGraphUpdate.bind(this);
         this.handlePublisherParameters = this.handlePublisherParameters.bind(this);
+        this.handleRowSelect = this.handleRowSelect.bind(this);
         this.getCurrentPage = this.getCurrentPage.bind(this);
+        this.getKey = this.getKey.bind(this);
         this.getUrlParameter = this.getUrlParameter.bind(this);
-        this.handleRowSelect = this.handleRowSelect.bind(this);
-
-
+        this.pageName = this.getCurrentPage();
     }
 
     static getProviderConf(widgetConfiguration) {
         return widgetConfiguration.configs.providerConfig;
     }
 
     handleRowSelect(event) {
-        //get the messageId from the selected row 
+        //get the messageId from the selected row
         let messageId = event.messageFlowId;
-        //substring and modify the URL, so on click, the page is directed to relevant message page with query parameter 'id'
-        let currPageStr = parent.window.location.href;
-        let index = currPageStr.lastIndexOf("/");
-        let tempMsgPageURL = currPageStr.substring(0, (index + 1)) + MESSAGE_PAGE;
-        let msgPageURL = new URL(tempMsgPageURL);
-        msgPageURL.searchParams.append(PARAM_ID, messageId);
-        window.location.href = msgPageURL;
+        let hashComponent = {};
+        hashComponent[this.getKey(MESSAGE_PAGE, "id")] = messageId;
+        window.location.href = MESSAGE_PAGE + ('#' + JSON.stringify(hashComponent));
```

Comment:
This will cause a full page refresh, which we should avoid in a React app.

Seems like this line save some data to the hash parameter in the URL.Why can't we use the `setWidgetState` in the base widget?

Reply:
This is not to set the widget state, but to send the state to another page. For that, we can't use the base class's setWidgetState method (it only changes the hash of the current page, but not the target). 

Since the widgets in the target page expect some of the data through the widget state (the object stored in the URL hash), we need to create that object from the current widget and passed that to the next page via the URL explicitly. That's why this has been done.

Reply:
OK I understand the problem. The issue with this approach is that it will cause a full page reload, which we should not do in a React app. 

IMO, we use the React Router in here. We can use `Redirect` (see [docs](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/docs/api/Redirect.md#to-object))
e.g. 
```js
<Redirect
    to={{
        pathname: '/uri-to-the-target-page',
        hash: 'your-hash-value',
        state: {messageId: 'msg-id-value'}
    }}
/>
```
A better approach will be, to pass the values through location state instead of hash param.
```
<Redirect
    to={{
        pathname: '/uri-to-the-target-page',
        state: {messageId: 'msg-id-value'}
    }}
/>
```
And in the target page, get `messageId` value as `this.props.location.state.messageId`